Update passkey-rs

iinuwa opened the pull request:
## 🎟️ Tracking

[PM-17241](https://bitwarden.atlassian.net/browse/PM-17241)

## 📔 Objective

This updates the SDK to point to a new version of our passkey-rs fork that has been synced with upstream. (The remaining differences between upstream and our fork are documented in the README of our fork.)

If approved, we should update `passkey-rs`'s `main` branch to point to the version referenced in this PR (https://github.com/bitwarden/passkey-rs/commit/2e5c0ba1fb60a3a6dec9c7dba175d3e07dc89862).

## 🚨 Breaking Changes

- `is_user_verification_enabled()` is no longer asynchronous. Clients should implement any dynamic checks for UV availability in the `check_user()` implementation.
- `user_handle has been added to the `find_credential()` method as an optional parameter. If the incoming request contains the user handle, then clients should pass it on. Otherwise, pass a `null` value, and the previous behavior.

iOS PR: https://github.com/bitwarden/ios/pull/2190

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-17241]: https://bitwarden.atlassian.net/browse/PM-17241?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

iinuwa committed:
fido: Bump to latest version of passkey-rs

iinuwa committed:
fido: Make user verification availability synchronous

This function is intended to signal whether user verification may potentially
be available for the authenticator, not whether user verification will be
available for a particular request. Since this can be known ahead of time for
our clients, there is no need for this function to be async. Making this async
also aligns us closer with upstream.

Per-request availability checks should be done in the caller's implementation
of check_user().

iinuwa committed:
fido: Add user handle to credential search parameters

dani-garcia commented:
I've noticed that your patches on top of 1Password's passkey-rs are bumping `rust-version = "1.85.1"` to `1.91.1`, is that really needed? That's putting a bound on the minimum version that the crate supports which is causing CI to fail on the SDK as we're pinning 1.91.0 for now.

Ideally that version should be as low as it can to increase compatibility, and none of the changes look like they need the bump.

iinuwa commented:
@dani-garcia, no, that's not required. I had seen the [recent bump in the SDK's Rust version to 1.91.1](https://github.com/bitwarden/sdk-internal/pull/576), so I assumed I should make this the same in our fork to unify the Rust versions across crates we owned. I can decrease it!

dani-garcia commented:
Ah, yeah updating the `rust-toolchain` file and the workflow YAML makes sense to match the versions we use in CI, but the `rust-version = ` field in the Cargo.toml doesn't indicate the version that will be used to build, but the minimum version that is allowed to build. So in that sense, the best option for us would be:
- `rust-toolchain` -> 1.91.1
- CI workflows -> 1.91.1
- `rust-version` in Cargo.toml -> 1.85.1 (for compatibility with older versions, even if we use latest on CI)

This ensures we use the latest version but allow users to build with older versions. You can see that we do the same thing ourselves, we bumped to 1.91.1, but our rust-version is still 1.85:
https://github.com/bitwarden/sdk-internal/blob/26df6719d3112171566d89f357fe4ce6e436ab49/Cargo.toml#L13

Sidenote, you may need to bump that file from 1.85 to 1.85.1 to match passkey-rs, otherwise CI might complain

dani-garcia commented:
Nice, thanks for the changes! Happy to know we're almost at the end goal of removing the need for our fork. Two more things that are causing problems in CI:

- You need to update the swift demo app, as it has a demo implementation of the fido2 traits:
https://github.com/bitwarden/sdk-internal/blob/4e5cb4ab7d62dcdf77e06dacaa1c6aa9292597bb/crates/bitwarden-uniffi/swift/iOS/App/ContentView.swift#L398-L428

- So apparently that minimum 1.85.1 version that the passkey-rs crate supports is not correct, since this let chain:
 https://github.com/bitwarden/passkey-rs/blob/873a4b983ef3fd7b24402a27da044739968ccbbd/passkey-client/src/extensions/prf.rs#L202-L203
is not supported until rust 1.88. This seems like an upstream problem that we're hitting because they're not testing their minimum supported rust version, so it might be worth reporting it upstream (so they can either remove the let-chain or bump their minimum rust version to 1.88)

iinuwa commented:
Thanks for pointing all that out. I think I've addressed all of the concerns now:
- I fixed the MSRV issue in our passkey-rs fork and created a PR to upstream: https://github.com/1Password/passkey-rs/pull/81
- I kept the MSRV in our fork as 1.85.1, but updated CI to use our standard Rust, Action, and OS runner versions.
- I updated the Swift demo.

iinuwa commented on crates/bitwarden-fido/src/types.rs:
```diff
@@ -472,9 +470,11 @@ impl TryFrom<UnverifiedAssetLink> for passkey::client::UnverifiedAssetLink<'_> {
+        let asset_link_url = {
+            let url = value
+                .asset_link_url
+                .unwrap_or_else(|| format!("https://{}/.well-known/assetlinks.json", value.host));
```
unwrap_or_else() will take the provided asset link URL or compute a fallback value from the value.host parameter. It is incorrect to say that this always generates a URL from the host.

This performs the same functionality as before: passkey-rs was doing this internally when None is passed, but has shifted that responsibility to the caller.

dani-garcia reviewed (approved):
LGTM, thanks!
